[pytorch] Multiprocessing api to use sigkill if sigterm doesn't kill the process

satgera opened the pull request:
Summary:
[pytorch] Multiprocessing api to use sigkill if sigterm doesn't kill the process
We have seen a handful of jobs training stuck where one of the trainer goes down
while others are stuck in c++ land and hence not handling the sigterm.

Test Plan: Manually validated by attaching gdb to one of the processes and sent a kill -9 to another. Saw the log ```WARNING] Unable to shutdown process 4422 via Signals.SIGTERM, forcefully exiting via Signals.SIGKILL```

Differential Revision: D51862545

wconstab commented on torch/multiprocessing/spawn.py:
```diff
@@ -126,9 +130,29 @@ def join(self, timeout=None):
+        # The reason is related to python signal handling explained
+        # below. We have seen processes getting stuck not handling
+        # SIGTERM.
+        # https://fb.workplace.com/groups/1543508939745889/permalink/1548605689236214/
```
nit: better to write out the explanation, as oss users can't read the workplace link.  (and its a pretty self-contained explanation anyway)

wconstab reviewed (approved):
lgtm. stamp to unblock. prefer to fix the nit.

tagging @zdevito @fduwjj to also take a 👀

fduwjj reviewed (approved):
LGTM as well.

satgera committed:
[pytorch] Multiprocessing api to use sigkill if sigterm doesn't kill the process (#115219)

Summary:

[pytorch] Multiprocessing api to use sigkill if sigterm doesn't kill the process
We have seen a handful of jobs training stuck where one of the trainer goes down
while others are stuck in c++ land and hence not handling the sigterm.

Test Plan: Manually validated by attaching gdb to one of the processes and sent a kill -9 to another. Saw the log ```WARNING] Unable to shutdown process 4422 via Signals.SIGTERM, forcefully exiting via Signals.SIGKILL```

Reviewed By: wconstab, fduwjj

Differential Revision: D51862545
